wolfcrypt/src/rsa.c

```diff
@@ -141,91 +145,101 @@ int wc_RsaFlattenPublicKey(RsaKey* key, byte* a, word32* aSz, byte* b,
     #include <wolfcrypt/src/misc.c>
 #endif
 
-#ifdef HAVE_CAVIUM
-    static int  InitCaviumRsaKey(RsaKey* key, void* heap);
-    static int  FreeCaviumRsaKey(RsaKey* key);
-    static int  CaviumRsaPublicEncrypt(const byte* in, word32 inLen, byte* out,
-                                       word32 outLen, RsaKey* key);
-    static int  CaviumRsaPrivateDecrypt(const byte* in, word32 inLen, byte* out,
-                                        word32 outLen, RsaKey* key);
-    static int  CaviumRsaSSL_Sign(const byte* in, word32 inLen, byte* out,
-                                  word32 outLen, RsaKey* key);
-    static int  CaviumRsaSSL_Verify(const byte* in, word32 inLen, byte* out,
-                                    word32 outLen, RsaKey* key);
-#endif
+#define ERROR_OUT(x) { ret = (x); goto done;}
 
-enum {
-    RSA_PUBLIC_ENCRYPT  = 0,
-    RSA_PUBLIC_DECRYPT  = 1,
-    RSA_PRIVATE_ENCRYPT = 2,
-    RSA_PRIVATE_DECRYPT = 3,
 
-    RSA_BLOCK_TYPE_1 = 1,
-    RSA_BLOCK_TYPE_2 = 2,
+#ifdef WOLFSSL_ASYNC_CRYPT
+    static int InitAsyncRsaKey(RsaKey* key);
+    static int FreeAsyncRsaKey(RsaKey* key);
+#endif /* WOLFSSL_ASYNC_CRYPT */
 
-    RSA_MIN_SIZE = 512,
-    RSA_MAX_SIZE = 4096,
+enum {
+    RSA_STATE_NONE = 0,
 
-    RSA_MIN_PAD_SZ   = 11      /* separator + 0 + pad value + 8 pads */
-};
+    RSA_STATE_ENCRYPT_PAD,
+    RSA_STATE_ENCRYPT_EXPTMOD,
+    RSA_STATE_ENCRYPT_RES,
 
+    RSA_STATE_DECRYPT_EXPTMOD,
+    RSA_STATE_DECRYPT_UNPAD,
+    RSA_STATE_DECRYPT_RES,
+};
 
-int wc_InitRsaKey(RsaKey* key, void* heap)
+int wc_InitRsaKey_ex(RsaKey* key, void* heap, int devId)
 {
-#ifdef HAVE_CAVIUM
-    if (key->magic == WOLFSSL_RSA_CAVIUM_MAGIC)
-        return InitCaviumRsaKey(key, heap);
-#endif
+    int ret = 0;
 
-    key->type = -1;  /* haven't decided yet */
-    key->heap = heap;
+    if (key == NULL) {
+        return BAD_FUNC_ARG;
+    }
 
-/* TomsFastMath doesn't use memory allocation */
-#ifndef USE_FAST_MATH
-    key->n.dp = key->e.dp = 0;  /* public  alloc parts */
+    (void)devId;
 
-    key->d.dp = key->p.dp  = 0;  /* private alloc parts */
-    key->q.dp = key->dP.dp = 0;
-    key->u.dp = key->dQ.dp = 0;
-#else
-    mp_init(&key->n);
-    mp_init(&key->e);
-    mp_init(&key->d);
-    mp_init(&key->p);
-    mp_init(&key->q);
-    mp_init(&key->dP);
-    mp_init(&key->dQ);
-    mp_init(&key->u);
+    key->type = RSA_TYPE_UNKNOWN;
+    key->state = RSA_STATE_NONE;
+    key->heap = heap;
+    key->tmp = NULL;
+    key->tmpLen = 0;
+
+#ifdef WOLFSSL_ASYNC_CRYPT
+    if (devId != INVALID_DEVID) {
+        /* handle as async */
+        ret = wolfAsync_DevCtxInit(&key->asyncDev, WOLFSSL_ASYNC_MARKER_RSA,
+                                                                        devId);
+        if (ret == 0) {
+            ret = InitAsyncRsaKey(key);
+        }
+    }
+    else
 #endif
+    {
+        mp_init(&key->n);
```

Comment:
Before we were only calling mp_init() in fastmath case so as to avoid getting memory until needed.  In normal math the pointers were set to 0 so there was a good state to cleanup or grow on demand.  Think of a server that may never use the peer RSA key 99% of the time or a client using ECC.  The wolfSSL_new() needs to init the RsaKey() before knowing so it's in a clean state.